donkeycar/parts/dgym.py

```diff
@@ -34,24 +35,46 @@ def __init__(self, sim_path, host="127.0.0.1", port=9091, headless=0, env_name="
                      'gyro': (0., 0., 0.),
                      'accel': (0., 0., 0.),
                      'vel': (0., 0., 0.)}
-        self.delay = float(delay)
+        self.delay = float(delay) / 1000
         self.record_location = record_location
         self.record_gyroaccel = record_gyroaccel
         self.record_velocity = record_velocity
         self.record_lidar = record_lidar
 
+        self.buffer = []
+
+    def delay_buffer(self, frame, info):
+        now = time.time()
+        buffer_tuple = (now, frame, info)
+        self.buffer.append(buffer_tuple)
+
+        # go through the buffer
+        num_to_remove = 0
+        for it in range(len(self.buffer)):
+            if now-self.buffer[it][0] >= self.delay:
+                num_to_remove += 1
+                self.frame = self.buffer[it][1]
+            else:
+                break
+
+        # clear the buffer
+        del self.buffer[:num_to_remove]
+
     def update(self):
         while self.running:
-            self.frame, _, _, self.info = self.env.step(self.action)
+            if self.delay > 0.0:
+                current_frame, _, _, current_info = self.env.step(self.action)
+                self.delay_buffer(current_frame, current_info)
+            else:
+                self.frame, _, _, self.info = self.env.step(self.action)
 
     def run_threaded(self, steering, throttle, brake=None):
         if steering is None or throttle is None:
             steering = 0.0
             throttle = 0.0
         if brake is None:
             brake = 0.0
-        if self.delay > 0.0:
-            time.sleep(self.delay / 1000.0)
```

Comment:
You are recalculating this number in every loop. Why don't you initialise `self.delay` already with the scaled number in the constructor in line 38?